### PR TITLE
Totp/#418 enable backend otp key storing

### LINKED ISSUE #418: Allow backend to store OTP keys and backup codes in the DB

### PR DESCRIPTION
**Pull request by kieranhalliday**
:tickets: **Ticket(s)**: Closes #418
---

## :construction_worker: Changes

Enabled totp storage, creation and deletion

## :flashlight: Testing Instructions

added TestEnableTOTP to users_test.go

### CODE REVIEW

**Comment by tristancalderbank** [comment]
^ should be fixed in https://github.com/ubclaunchpad/inertia/pull/434, I think that just needs to be merged then kieran can rebase

**Comment by bobheadxi** [comment]
I've merged #434

File: daemon/inertiad/auth/users.go [hunk 1]
```diff
@@ -20,9 +20,11 @@ const (
 // userProps are properties associated with user, used
 // for database entries
 type userProps struct {
-	HashedPassword string
-	Admin          bool
-	LoginAttempts  int
+	HashedPassword  string
+	Admin           bool
+	LoginAttempts   int
+	totpKey         string
```
Comment (bfbachmann) [inline daemon/inertiad/auth/users.go]: Shouldn't this be exported?

**Comment by bfbachmann** [comment]
@bobheadxi We're planning on updating this to encrypt TOTP keys and backup codes once @tristancalderbank adds the required crypto libraries, so hold off on review for now.

File: daemon/inertiad/auth/users.go (hunk 1 above)
Comment (tristancalderbank) [reply daemon/inertiad/auth/users.go]: This is also going to need a totpSalt key as PBKDF2 requires a salt

File: daemon/inertiad/auth/users.go [hunk 2]
```diff
@@ -227,3 +229,87 @@ func (m *userManager) IsAdmin(username string) (bool, error) {
...
+	return err
+}
+
+// disableTOTP disables TOTP for a user
+func (m *userManager) DisableTOTP(username string) error {
+
+	err := m.db.Update(func(tx *bolt.Tx) error {
+		users := tx.Bucket(m.usersBucket)
+		propsBytes := users.Get([]byte(username))
+		if propsBytes != nil {
```
Comment (bfbachmann) [inline daemon/inertiad/auth/users.go]: I think my comment from above about the user not existing applies here too.

File: daemon/inertiad/crypto/key_derivation.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+// GenerateSalt returns a random hex encoded salt for KD algorithm
+func GenerateSalt() []byte {
+	salt := make([]byte, KeyDerivationSaltLength)
+	rand.Read(salt)
+	return salt
+}
+
+// DeriveKey derives an AES encryption key based on salt + user's
+// password using PBKDF2 with HMAC-SHA1
```
Comment (bfbachmann) [inline daemon/inertiad/crypto/key_derivation.go]: I think you meant to say `HMAC-SHA256` here, since that's what you're using in the function.

**Comment by bfbachmann** [comment]
I'm closing this because it's commit history is botched. There's a new PR up here: https://github.com/ubclaunchpad/inertia/pull/449.

File: daemon/inertiad/crypto/key_derivation.go (hunk 3 above)
Comment (kieranhalliday) [reply daemon/inertiad/crypto/key_derivation.go]: Are you sure? This is Tristan's work but look at line 15 in totp.go

**Pull request closed without merging**